File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/ui/base/DokitViewManager.java
package com.didichuxing.doraemonkit.ui.base;

import android.app.Activity;
import android.arch.persistence.room.Room;
import android.content.Context;
import android.graphics.Point;
import android.view.WindowManager;

import com.didichuxing.doraemonkit.constant.DokitConstant;
import com.didichuxing.doraemonkit.kit.network.room_db.DokitDatabase;
import com.didichuxing.doraemonkit.kit.network.room_db.DokitDbManager;
import com.didichuxing.doraemonkit.ui.main.ToolPanelDokitView;
import com.didichuxing.doraemonkit.util.LogHelper;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by jintai on 2018/10/23.
 * 浮标管理类
 */

public class DokitViewManager implements DokitViewManagerInterface {
    public static final String TAG = "DokitViewManagerProxy";
    /**
     * 每个类型在页面中的位置 只保存marginLeft 和marginTop
     */
    private static Map<String, Point> mDokitViewPos;

    private Map<String, LastDokitViewPosInfo> mLastDokitViewPosInfoMaps;


    private DokitViewManagerInterface mDokitViewManager;
    private Context mContext;
    /**
     * 数据库操作类
     */
    private DokitDatabase mDB;

    /**
     * 静态内部类单例
     */
    private static class Holder {
        private static DokitViewManager INSTANCE = new DokitViewManager();
    }

    public static DokitViewManager getInstance() {
        return DokitViewManager.Holder.INSTANCE;
    }


    public void init(Context context) {
        mContext = context;
        if (DokitConstant.IS_NORMAL_FLOAT_MODE) {
            mDokitViewManager = new NormalDokitViewManager(context);
        } else {
            mDokitViewManager = new SystemDokitViewManager(context);
        }
        mDokitViewPos = new HashMap<>();
        mLastDokitViewPosInfoMaps = new HashMap<>();
        mDB = Room.databaseBuilder(context,
                DokitDatabase.class,
                "dokit-database")
                //下面注释表示允许主线程进行数据库操作，但是不推荐这样做。
                //他可能造成主线程lock以及anr
                //所以我们的操作都是在新线程完成的
                .allowMainThreadQueries()
                .build();
        //获取所有的intercept apis
        DokitDbManager.getInstance().getAllInterceptApis();

        //获取所有的template apis
        DokitDbManager.getInstance().getAllTemplateApis();

    }

    public DokitDatabase getDb() {
        return mDB;
    }

    /**
     * 当app进入后台时调用
     */
    @Override
    public void notifyBackground() {
        mDokitViewManager.notifyBackground();
    }

    /**
     * 当app进入前台时调用
     */
    @Override
    public void notifyForeground() {
        mDokitViewManager.notifyForeground();
    }

    /**
     * 只有普通浮标才会调用
     * 保存每种类型dokitView的位置
     */
    public void saveDokitViewPos(String tag, int marginLeft, int marginTop) {
        if (mDokitViewPos == null) {
            return;
        }

        if (mDokitViewPos.get(tag) == null) {
            Point point = new Point(marginLeft, marginTop);
            mDokitViewPos.put(tag, point);
        } else {
            Point point = mDokitViewPos.get(tag);
            point.set(marginLeft, marginTop);
        }

        for (String key : mDokitViewPos.keySet()) {
            LogHelper.i(TAG, "saveDokitViewPos  key==> " + key + "  point===>" + mDokitViewPos.get(key));
        }
    }

    /**
     * 只有普通的浮标才需要调用
     * 获得指定dokitView的位置信息
     *
     * @param tag
     * @return
     */
    Point getDokitViewPos(String tag) {
        if (mDokitViewPos == null) {
            return null;
        }

        for (String key : mDokitViewPos.keySet()) {
            LogHelper.i(TAG, "getDokitViewPos  key==> " + key + "  point===>" + mDokitViewPos.get(key));
        }
        return mDokitViewPos.get(tag);
    }

    /**
     * 只有普通的浮标才需要调用
     * 添加activity关联的所有dokitView activity resume的时候回调
     *
     * @param activity
     */
    @Override
    public void resumeAndAttachDokitViews(Activity activity) {
        mDokitViewManager.resumeAndAttachDokitViews(activity);
    }

    @Override
    public void onMainActivityCreate(Activity activity) {

    }

    @Override
    public void onActivityCreate(Activity activity) {

    }

    @Override
    public void onActivityResume(Activity activity) {

    }

    /**
     * 在当前Activity中添加指定悬浮窗
     *
     * @param dokitIntent
     */
    @Override
    public void attach(DokitIntent dokitIntent) {
        mDokitViewManager.attach(dokitIntent);
    }

    /**
     * 隐藏工具列表dokitView
     */
    public void detachToolPanel() {
        detach(ToolPanelDokitView.class.getSimpleName());
    }

    /**
     * 移除每个activity指定的dokitView
     */
    @Override
    public void detach(String tag) {
        mDokitViewManager.detach(tag);
    }


    /**
     * 移除每个activity指定的dokitView
     */
    @Override
    public void detach(AbsDokitView dokitView) {
        mDokitViewManager.detach(dokitView);
    }

    /**
     * 移除所有activity的所有dokitView
     */
    @Override
    public void detachAll() {
        mDokitViewManager.detachAll();
    }

    @Override
    public void detach(Class<? extends AbsDokitView> dokitViewClass) {
        mDokitViewManager.detach(dokitViewClass);
    }


    /**
     * Activity销毁时调用
     */
    @Override
    public void onActivityDestroy(Activity activity) {
        mDokitViewManager.onActivityDestroy(activity);
    }

    /**
     * 获取页面上指定的dokitView
     *
     * @param activity 如果是系统浮标 activity可以为null
     * @param tag
     * @return
     */
    @Override
    public AbsDokitView getDokitView(Activity activity, String tag) {
        return mDokitViewManager.getDokitView(activity, tag);
    }

    /**
     * @param activity
     * @return
     */
    @Override
    public Map<String, AbsDokitView> getDokitViews(Activity activity) {
        return mDokitViewManager.getDokitViews(activity);
    }

    /**
     * 系统悬浮窗需要调用
     */
    public interface DokitViewAttachedListener {
        void onDokitViewAdd(AbsDokitView dokitView);
    }

    /**
     * 系统悬浮窗需要调用
     *
     * @param listener
     */
    void addDokitViewAttachedListener(DokitViewAttachedListener listener) {
        if (!DokitConstant.IS_NORMAL_FLOAT_MODE && mDokitViewManager instanceof SystemDokitViewManager) {
            ((SystemDokitViewManager) mDokitViewManager).addListener(listener);
        }
    }

    /**
     * 系统悬浮窗需要调用
     *
     * @param listener
     */
    void removeDokitViewAttachedListener(DokitViewAttachedListener listener) {
        if (!DokitConstant.IS_NORMAL_FLOAT_MODE && mDokitViewManager instanceof SystemDokitViewManager) {
            ((SystemDokitViewManager) mDokitViewManager).removeListener(listener);
        }
    }

    /**
     * 获取
     *
     * @return WindowManager
     */
    WindowManager getWindowManager() {
        return (WindowManager) mContext.getSystemService(Context.WINDOW_SERVICE);
    }

    void saveLastDokitViewPosInfo(String key, LastDokitViewPosInfo lastDokitViewPosInfo) {
        if (mLastDokitViewPosInfoMaps != null) {
            mLastDokitViewPosInfoMaps.put(key, lastDokitViewPosInfo);
        }
    }

    LastDokitViewPosInfo getLastDokitViewPosInfo(String key) {
        if (mLastDokitViewPosInfoMaps == null) {
            return null;
        }
        return mLastDokitViewPosInfoMaps.get(key);
    }


    void removeLastDokitViewPosInfo(String key) {
        if (mLastDokitViewPosInfoMaps == null) {
            return;
        }
        mLastDokitViewPosInfoMaps.remove(key);
    }

}

File: Android/dokit/src/main/java/com/didichuxing/doraemonkit/kit/core/DokitViewManager.kt
package com.didichuxing.doraemonkit.kit.core

import android.app.Activity
import android.content.Context
import android.content.res.Configuration
import android.graphics.Point
import android.view.WindowManager
import androidx.collection.ArrayMap
import androidx.room.Room
import com.didichuxing.doraemonkit.DoKitEnv
import com.didichuxing.doraemonkit.kit.network.room_db.DokitDatabase
import com.didichuxing.doraemonkit.kit.network.room_db.DokitDbManager
import com.didichuxing.doraemonkit.util.ScreenUtils

/**
 * Created by jintai on 2018/10/23.
 * 浮标管理类
 */
class DokitViewManager : DokitViewManagerInterface {

    companion object {
        private const val TAG = "DokitViewManagerProxy"

        @JvmStatic
        val INSTANCE: DokitViewManager by lazy { DokitViewManager() }

        /**
         * 每个类型在页面中的位置 只保存marginLeft 和marginTop
         */
        private val doKitViewPos: MutableMap<String, DoKitViewInfo> = ArrayMap<String, DoKitViewInfo>()
    }

    /**
     * Retrieves app [WindowManager]
     *
     * @return WindowManager
     */
    val windowManager: WindowManager
        get() = DoKitEnv.requireApp().getSystemService(Context.WINDOW_SERVICE) as WindowManager

    private val lastDoKitViewPosInfoMaps: MutableMap<String, LastDokitViewPosInfo> = ArrayMap<String, LastDokitViewPosInfo>()

    private val listeners: MutableList<DokitViewManager.DokitViewAttachedListener> by lazy { mutableListOf<DokitViewManager.DokitViewAttachedListener>() }

    private var _doKitViewManager: AbsDokitViewManager? = null

    //下面注释表示允许主线程进行数据库操作，但是不推荐这样做。
    //他可能造成主线程lock以及anr
    //所以我们的操作都是在新线程完成的
    val db: DokitDatabase by lazy {
        Room.databaseBuilder(
            DoKitEnv.requireApp(),
            DokitDatabase::class.java,
            "dokit-database"
        ) //下面注释表示允许主线程进行数据库操作，但是不推荐这样做。
            //他可能造成主线程lock以及anr
            //所以我们的操作都是在新线程完成的
            .allowMainThreadQueries()
            .build()
    }

    fun init() {
        //获取所有的intercept apis
        DokitDbManager.getInstance().getAllInterceptApis()

        //获取所有的template apis
        DokitDbManager.getInstance().getAllTemplateApis()
    }

    /**
     * 当app进入后台时调用
     */
    override fun notifyBackground() {
        ensureViewManager().notifyBackground()
    }

    /**
     * 当app进入前台时调用
     */
    override fun notifyForeground() {
        ensureViewManager().notifyForeground()
    }

    /**
     * 只有普通浮标才会调用
     * 保存每种类型dokitView的位置
     */
    fun saveDokitViewPos(tag: String, marginLeft: Int, marginTop: Int) {

        var orientation = -1
        val portraitPoint = Point()
        val landscapePoint = Point()
        if (ScreenUtils.isPortrait()) {
            orientation = Configuration.ORIENTATION_PORTRAIT
            portraitPoint.x = marginLeft
            portraitPoint.y = marginTop
        } else {
            orientation = Configuration.ORIENTATION_LANDSCAPE
            landscapePoint.x = marginLeft
            landscapePoint.y = marginTop
        }
        if (doKitViewPos[tag] == null) {
            val doKitViewInfo = DoKitViewInfo(orientation, portraitPoint, landscapePoint)
            doKitViewPos[tag] =
                doKitViewInfo
        } else {
            val doKitViewInfo = doKitViewPos[tag]
            if (doKitViewInfo != null) {
                doKitViewInfo.orientation = orientation
                doKitViewInfo.portraitPoint = portraitPoint
                doKitViewInfo.landscapePoint = landscapePoint
            }
        }
    }

    /**
     * 只有普通的浮标才需要调用
     * 获得指定dokitView的位置信息
     *
     * @param tag
     * @return
     */
    fun getDoKitViewPos(tag: String): DoKitViewInfo? = doKitViewPos[tag]

    /**
     * 只有普通的浮标才需要调用
     * 添加activity关联的所有dokitView activity resume的时候回调
     *
     * @param activity
     */
    override fun dispatchOnActivityResumed(activity: Activity?) {
        activity?.also { ensureViewManager().dispatchOnActivityResumed(it) }
    }

    override fun onActivityPaused(activity: Activity?) {
        activity?.also { ensureViewManager().onActivityPaused(it) }
    }

    override fun onActivityStopped(activity: Activity?) {
        activity?.also { ensureViewManager().onActivityStopped(it) }
    }

    /**
     * 在当前Activity中添加指定悬浮窗
     *
     * @param dokitIntent
     */
    override fun attach(dokitIntent: DokitIntent) {
        ensureViewManager().attach(dokitIntent)
    }

    /**
     * 隐藏工具列表dokitView
     */
    fun detachToolPanel() {
        ensureViewManager().detachToolPanel()
    }

    /**
     * 显示工具列表dokitView
     */
    fun attachToolPanel(activity: Activity) {
        ensureViewManager().attachToolPanel(activity)
    }

    /**
     * 显示主图标 dokitView
     */
    fun attachMainIcon(activity: Activity) {
        ensureViewManager().attachMainIcon(activity)
    }

    /**
     * 隐藏首页图标
     */
    fun detachMainIcon() {
        ensureViewManager().detachMainIcon()
    }

    /**
     * 移除每个activity指定的dokitView
     */
    override fun detach(tag: String) {
        ensureViewManager().detach(tag)
    }

    /**
     * 移除每个activity指定的dokitView
     */
    override fun detach(dokitView: AbsDokitView) {
        ensureViewManager().detach(dokitView)
    }

    override fun detach(doKitViewClass: Class<out AbsDokitView>) {
        ensureViewManager().detach(doKitViewClass)
    }

    /**
     * 移除所有activity的所有dokitView
     */
    override fun detachAll() {
        ensureViewManager().detachAll()
    }

    /**
     * 获取页面上指定的dokitView
     *
     * @param activity 如果是系统浮标 activity可以为null
     * @param tag
     * @return
     */
    override fun <T : AbsDokitView> getDoKitView(activity: Activity?, clazz: Class<T>): AbsDokitView? {
        return activity?.let { ensureViewManager().getDoKitView(it, clazz) }
    }

    /**
     * Activity销毁时调用
     */
    override fun onActivityDestroyed(activity: Activity?) {
        activity?.also { ensureViewManager().onActivityDestroyed(it) }
    }

    /**
     * @param activity
     * @return
     */
    override fun getDoKitViews(activity: Activity?): Map<String, AbsDokitView>? {
        return activity?.let { ensureViewManager().getDoKitViews(it) }
    }

    /**
     * DokitView浮窗添加后的回调
     */
    interface DokitViewAttachedListener {
        fun onDokitViewAdd(dokitView: AbsDokitView?)
    }

    /**
     * 系统悬浮窗需要调用
     *
     * @param listener
     */
    fun addDokitViewAttachedListener(listener: DokitViewAttachedListener?) {
        listener?.let{
            listeners.add(listener)
        }
    }

    /**
     * 系统悬浮窗需要调用
     *
     * @param listener
     */
    fun removeDokitViewAttachedListener(listener: DokitViewAttachedListener?) {
        listener?.let{
            listeners.remove(listener)
        }
    }

    fun saveLastDokitViewPosInfo(key: String, lastDokitViewPosInfo: LastDokitViewPosInfo) {
        lastDoKitViewPosInfoMaps[key] = lastDokitViewPosInfo
    }

    fun getLastDokitViewPosInfo(key: String): LastDokitViewPosInfo? = lastDoKitViewPosInfoMaps[key]

    fun removeLastDokitViewPosInfo(key: String) {
        lastDoKitViewPosInfoMaps.remove(key)
    }

    @Synchronized
    private fun ensureViewManager(): AbsDokitViewManager {
        return _doKitViewManager
            ?: run {
                if (DoKitManager.IS_NORMAL_FLOAT_MODE) NormalDoKitViewManager() else SystemDoKitViewManager()
            }.also {
                _doKitViewManager = it
            }
    }

    internal fun notifyDokitViewAdd(dokitView: AbsDokitView?) {
        for (listener in listeners) {
            listener.onDokitViewAdd(dokitView)
        }
    }
}
